python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,491 @@
+        if self.groupFailure(testers):
+            return
+
+        # Local storage of skipped tester objects
+        skipped_or_failed_tests = set([])
+
+        # Local DAG outside the tester loop (one DAG for entire group)
+        tester_dag = dag.DAG()
+
+        # Create the DAG, skipped_testers, and just about anything we can during this first iteration loop
+        for tester in testers:
+
+            # Local tester_data object to contain individual tester. The TesterData object is responsible
+            # for handling the subprocess object and the output file handler.
+            tester_data = TesterData(tester, tester_dag, self.options)
+
+            with self.thread_lock:
+                # Create tester_data map for each tester. This will contain and allow access to the the tester,
+                # and the associated DAG for the entire tester group.
+                self.scheduled_tester_data[tester.getTestName()] = tester_data
+
+            # Create a node for this tester if not already exists (prereq iteration creates nodes)
+            if not tester_dag.node_exists(tester.getTestName()):
+                tester_dag.add_node(tester.getTestName())
+
+            # Loop through prereqs and add leaves/nodes where necessary
+            for prereq in tester.getPrereqs():
+
+                # Discover unknown prereqs
+                if prereq not in [x.getTestName() for x in testers]:
+                    tester.setStatus('unknown dependency', tester.bucket_fail)
+                    skipped_or_failed_tests.add(tester_data)
+                    continue
+
+                if not tester_dag.node_exists(prereq):
+                    # Create a node for this tester
+                    tester_dag.add_node(prereq)
+
+                # Once you add an edge to a node, the DAG arranges the nodes accordingly
+                try:
+                    tester_dag.add_edge(prereq, tester.getTestName())
+
+                # Handle Cyclic errors
+                except dag.DAGValidationError:
+                    tester.setStatus('Cyclic or Invalid Dependency Detected!', tester.bucket_fail)
+                    skipped_or_failed_tests.add(tester_data)
+
+                    # Pass so we allow the status runner to print the results serially
+                    pass
+
+            # Discover skipped tests
+            if not tester.getRunnable(self.options):
+                skipped_or_failed_tests.add(tester_data)
+
+        # Discover skipped tests and remove downstream tests that require this skipped test. Any
+        # discovered downstream nodes will have 'skipped dependency' set as their status.
+        for skipped_tester_data in skipped_or_failed_tests:
+            tmp_tester = skipped_tester_data.getTester()
+
+            # Delete downstream nodes
+            self.deleteDownstreamTests(skipped_tester_data)
+
+            # Delete this node
+            tester_dag.delete_node_if_exists(tmp_tester.getTestName())
```

Comment:
Your new method which you will create `delete_sub_tree()` should probably delete the whole subtree rooted at the given node. If you do that, you won't have to delete the downstream, then the current node. It'll all be one operation.